client/src/actors.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use riker::actors::*;
+
+use std::{fmt::Debug, path::PathBuf};
+
+use engine::vault::{BoxProvider, Key, RecordHint, RecordId};
+
+use crate::{
+    bucket::Bucket,
+    client::{ClientMsg, InternalResults},
+    ids::VaultId,
+    key_store::KeyStore,
+    line_error,
+    provider::Provider,
+    snapshot::Snapshot,
+};
+
+/// Messages used to talk to the Bucket Actor.
+#[derive(Debug, Clone)]
+pub enum BMsg<P: BoxProvider + Debug> {
+    CreateVault(VaultId, Key<P>),
+    ReadData(Key<P>, RecordId),
+    WriteData(Key<P>, RecordId, Vec<u8>, RecordHint),
+    InitRecord(Key<P>, VaultId),
+    RevokeData(Key<P>, RecordId),
+    GarbageCollect(Key<P>),
+    ListAsk(Key<P>),
+    WriteSnapshot(String, Option<String>, Option<PathBuf>),
+    ReadSnapshot(String, Option<String>, Option<PathBuf>),
+    ReloadData(Vec<u8>),
+}
+
+/// Messages used for the KeyStore Actor.
+#[derive(Clone, Debug)]
+pub enum KMsg {
+    CreateVault(VaultId),
+    ReadData(VaultId, RecordId),
+    WriteData(VaultId, RecordId, Vec<u8>, RecordHint),
+    InitRecord(VaultId),
+    RevokeData(VaultId, RecordId),
+    GarbageCollect(VaultId),
+    ListIds(VaultId),
+    RebuildKeys(Vec<Key<Provider>>, Vec<Vec<RecordId>>),
+}
+
+/// Messages used for the Snapshot Actor.
+#[derive(Clone, Debug)]
+pub enum SMsg {
+    WriteSnapshot(String, Option<String>, Option<PathBuf>, Vec<u8>),
+    ReadSnapshot(String, Option<String>, Option<PathBuf>),
+}
+
+/// Actor Factory for the Bucket.
+impl ActorFactory for Bucket<Provider> {
+    fn create() -> Self {
+        Bucket::new()
+    }
+}
+
+/// Actor Factory for the KeyStore.
+impl ActorFactory for KeyStore<Provider> {
+    fn create() -> Self {
+        KeyStore::new()
+    }
+}
+
+/// Actor Factory for the Snapshot.
+impl ActorFactory for Snapshot {
+    fn create() -> Self {
+        Snapshot::new::<Provider>(vec![])
+    }
+}
+
+impl Actor for Bucket<Provider> {
+    type Msg = BMsg<Provider>;
+
+    fn recv(&mut self, ctx: &Context<Self::Msg>, msg: Self::Msg, sender: Sender) {
+        self.receive(ctx, msg, sender);
+    }
+}
+
+impl Actor for KeyStore<Provider> {
+    type Msg = KMsg;
+
+    fn recv(&mut self, ctx: &Context<Self::Msg>, msg: Self::Msg, sender: Sender) {
+        self.receive(ctx, msg, sender);
+    }
+}
+
+impl Actor for Snapshot {
+    type Msg = SMsg;
+
+    fn recv(&mut self, ctx: &Context<Self::Msg>, msg: Self::Msg, sender: Sender) {
+        self.receive(ctx, msg, sender);
+    }
+}
+
+impl Receive<SMsg> for Snapshot {
+    type Msg = SMsg;
+
+    fn receive(&mut self, ctx: &Context<Self::Msg>, msg: Self::Msg, _sender: Sender) {
+        match msg {
+            SMsg::WriteSnapshot(pass, name, path, state) => {
```

Comment:
does this mean I need to manually tell stronghold when I want to save the pending changes to a snapshot file?

Reply:
For now, there is no periodic "write" job.  However, that is something on the priority list.  To add functionality to allow you to specify that duration.  I also may add some features to allow the writes to happen after every single change to the state of the snapshot.